### PR TITLE
Add X25519Kyber768Draft00 experimental HPKE KEM

### PR DESCRIPTION
**Pull request by chris-wood**
See [the draft](https://bwesterb.github.io/draft-westerbaan-cfrg-hpke-xyber768d00/draft-westerbaan-cfrg-hpke-xyber768d00.html). Some tasks still left to do:

- [x] Add multiple encryption vectors to match RFC9180
- [x] Add multiple export vectors to match RFC9180

### CODE REVIEW

**Review by bwesterb — CHANGES_REQUESTED** [review]

**Comment by bwesterb** [comment]
Thanks @chris-wood !

**Comment by bwesterb** [comment]
For your convenience:

```
  Running [/home/runner/golangci-lint-1.51.2-linux-amd64/golangci-lint run --out-format=github-actions --config=./.etc/golangci.yml ./...] in [] ...
  Error: File is not `gofmt`-ed with `-s` (gofmt)
  Error: File is not `gofmt`-ed with `-s` (gofmt)
  Error: File is not `gofmt`-ed with `-s` (gofmt)
  Error: File is not `gofumpt`-ed (gofumpt)
  Error: File is not `gofumpt`-ed (gofumpt)
  Error: `(kemBase).labeledExtract` - `salt` always receives `nil` (unparam)
  ```

File: hpke/hybridkem.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+)
+
+type hybridKEM struct {
+	kemBase
+	kemA kem.Scheme
+	kemB kem.Scheme
+}
+
+func (h hybridKEM) PrivateKeySize() int        { return h.kemA.PrivateKeySize() + h.kemB.PrivateKeySize() }
+func (h hybridKEM) SeedSize() int              { return h.kemA.SeedSize() + h.kemB.SeedSize() }
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hybridkem.go:19]: Does it though? Looking [here](https://bwesterb.github.io/draft-westerbaan-cfrg-hpke-xyber768d00/draft-westerbaan-cfrg-hpke-xyber768d00.html#name-derivekeypair), it seems to be 32 + 64. Am I looking at the wrong thing?

File: hpke/hybridkem.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+func (h hybridKEM) AuthEncapsulateDeterministically(pkr kem.PublicKey, sks kem.PrivateKey, seed []byte) (ct, ss []byte, err error) {
+	panic("AuthEncapsulateDeterministically is not supported for this KEM")
+}
+
+func (h hybridKEM) Encapsulate(pkr kem.PublicKey) (
+	ct []byte, ss []byte, err error,
+) {
+	panic("Encapsulate is not implemented")
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hybridkem.go:46]: I don't think it's used by HPKE, so I don't think we need it.

File: hpke/hybridkem.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+func (k *hybridKEMPrivKey) Public() kem.PublicKey {
+	return &hybridKEMPubKey{
+		scheme: k.scheme,
+		pubA: k.privA.Public(),
+		pubB: k.privB.Public(),
+	}
+}
+
+func (k *hybridKEMPrivKey) Validate() bool {
+	// Remove once the HPKE caller is gone
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hybridkem.go:127]: As discussed offline, validation would happen upon decoding. This comment suggests that Validate() as a function isn't really needed. I suggest we make that change separately.

File: hpke/hybridkem.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+func (k *hybridKEMPubKey) Equal(pk kem.PublicKey) bool {
+	k1, ok := pk.(*hybridKEMPubKey)
+	return ok &&
+		k.pubA.Equal(k1.pubA) &&
+		k.pubB.Equal(k1.pubB)
+}
+
+func (k *hybridKEMPubKey) Validate() bool {
+	// Remove once the HPKE caller is gone
+	return true
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hybridkem.go:167]: Same here.

File: hpke/hybridkem.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	if len(seed) != h.SeedSize() {
+		panic(kem.ErrSeedSize)
+	}
+
+	dkpPrk := h.labeledExtract(nil, []byte("dkp_prk"), seed)
+	bytes := h.labeledExpand(
+		dkpPrk,
+		[]byte("sk"),
+		nil,
+		uint16(h.SeedSize()),
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hybridkem.go:187]: It is! That's how this function is implemented.

File: hpke/hybridkem.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	pubKey := &hybridKEMPubKey{
+		pubA: pubA,
+		pubB: pubB,
+	}
+
+
+	return pubKey, privKey
+}
+
+func (h hybridKEM) GenerateKeyPair() (kem.PublicKey, kem.PrivateKey, error) {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hybridkem.go:207]: I'm fine with that, too.

File: hpke/hybridkem.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
+package hpke
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hybridkem.go:1]: Would you be OK with just copying that text?

**Comment by chris-wood** [comment]
> For your convenience:
> 
> ```
>   Running [/home/runner/golangci-lint-1.51.2-linux-amd64/golangci-lint run --out-format=github-actions --config=./.etc/golangci.yml ./...] in [] ...
>   Error: File is not `gofmt`-ed with `-s` (gofmt)
>   Error: File is not `gofmt`-ed with `-s` (gofmt)
>   Error: File is not `gofmt`-ed with `-s` (gofmt)
>   Error: File is not `gofumpt`-ed (gofumpt)
>   Error: File is not `gofumpt`-ed (gofumpt)
>   Error: `(kemBase).labeledExtract` - `salt` always receives `nil` (unparam)
> ```

What incantation do I need to invoke to make these changes locally? Can we get that set up in a `make` target?

File: hpke/hybridkem.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bwesterb) [reply hpke/hybridkem.go:21]: It stretches a 32 byte preseed to a 32 and a 64 byte seed.

File: hpke/hybridkem.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bwesterb) [reply hpke/hybridkem.go:129]: Ok! But then usage is blocked until we make that change as well.

File: hpke/hybridkem.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hybridkem.go]: Ah! Yeah, `ikm` is 32 bytes. Got it. I fixed that.

**Comment by bwesterb** [comment]
> What incantation do I need to invoke to make these changes locally? Can we get that set up in a `make` target?

`make lint`

**Comment by chris-wood** [comment]
> > What incantation do I need to invoke to make these changes locally? Can we get that set up in a `make` target?
> 
> `make lint`

This is sort of unhelpful in that it just runs the linter without actually applying changes. I'll send a PR to update the Makefile so that there's a target for actually doing the formatting.

**Comment by chris-wood** [comment]
@bwesterb I pushed some preliminary test vector generation code.

**Comment by chris-wood** [comment]
Marking as ready for review now. I updated when keys are validated (only when we receive them over the wire through deserialization functions). Fail early and fail often!

**Review by bwesterb — APPROVED** [review]

File: hpke/hybrid-x25119-kyber768-test-vectors.json [hunk 8]
```diff
@@ -0,0 +1 @@
...8cde901f296f0c050e9586bd5a7f12f8f087cab9d2164ac41102062619a314639c051787e5dff24896c7e9ed7fe299a598fabc9dd15671e26e744a514ca6b0328235c919def4ad8a835e87f41519a661d1fc8f59c85f3c9ca8544325ca0932abc6bea33f47e43a84e1d","shared_secret":"6cf976165667e7da5d06ddbb7840f8d766609ee1a3f3bd1c947a4498afb4c4ac9137e2e99a3bb83da5c01108f30150e33c804300c3032d212a0f63c2d6bdd631","key_schedule_context":"009f749a195d1c8b3eaa8d5c3f571dc7231aafbbc0405e4b484738352667c4848677d1a0316e6519bc5ce5bb9ab1eaad121d47115eeecde2c60abef07a70ada444","secret":"eb07f25547b772c34e6258b21be0c29bd8d342fc1c1971a71ee8adea30ddd678","key":"8bfee5330f68713fc27fb74d4876f97c","base_nonce":"99c08a3cda9f6feb763dcec8","exporter_secret":"850da245cff2a8bc02f1dc2f6ea66bea376c92eba09a47a89e392abfe6a91619","encryptions":[{"aad":"6164646974696f6e616c207075626c69632064617461","ct":"c2c92b8c097ed475b0e26fa4e409ce570965c4eb6d7804ecfdfe423f1dc9c9949cc15b8d50e1aa9a1e6fddebe814b2ea8834","nonce":"99c08a3cda9f6feb763dcec8","pt":"7465787420656e6372797074656420746f20426f622773207075626c6963206b6579"}],"exports":[{"exporter_context":"7465737420636f6e74657874","L":32,"exported_value":"4824948ac0b9d9a47c4ee80d3d0269006bdbe18f637df9d760fa77b1ca19a023"}]}]
```
Comment (bwesterb) [inline hpke/hybrid-x25119-kyber768-test-vectors.json:1]: Did you intend to commit this file?

Comment (chris-wood) [reply hpke/hybrid-x25119-kyber768-test-vectors.json:1]: Yeah, but we can pull it out if we don't want it in the repo.

**Comment by bwesterb** [comment]
@armfazh @cjpatton Please check whether this makes sense to you.

File: hpke/hpke.go [hunk 9]
```diff
@@ -105,10 +105,6 @@ type Sender struct {
 
 // NewSender creates a Sender with knowledge of the receiver's public-key.
 func (suite Suite) NewSender(pkR kem.PublicKey, info []byte) (*Sender, error) {
-	if !suite.kemID.validatePublicKey(pkR) {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/hpke.go:108]: Let's address that in a separate PR?

File: hpke/aead.go [hunk 10]
```diff
@@ -7,11 +7,14 @@ import (
 
 type encdecContext struct {
 	// Serialized parameters
-	suite          Suite
-	exporterSecret []byte
-	key            []byte
-	baseNonce      []byte
-	sequenceNumber []byte
+	suite              Suite
+	sharedSecret       []byte
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke/aead.go:11]: Yes, these are intermediate values needed when producing test vectors.

**Comment by chris-wood** [comment]
@armfazh squashed and ready to go!

**Comment by bwesterb** [comment]
Cool 😎

**Pull request merged**